File: packages/python/qraudio/profiles.py
from __future__ import annotations

from enum import Enum
from typing import Iterable


class ProfileName(str, Enum):
    AFSK_BELL = "afsk-bell"
    AFSK_FIFTH = "afsk-fifth"
    GFSK_FIFTH = "gfsk-fifth"
    MFSK = "mfsk"


Profile = ProfileName

PROFILE_NAMES: tuple[ProfileName, ...] = (
    ProfileName.AFSK_BELL,
    ProfileName.AFSK_FIFTH,
    ProfileName.GFSK_FIFTH,
    ProfileName.MFSK,
)

DEFAULT_PROFILE: Profile = ProfileName.AFSK_BELL

_PROFILE_SET = set(PROFILE_NAMES)


def isProfile(value: object) -> bool:
    if isinstance(value, ProfileName):
        return True
    if isinstance(value, str):
        return value in _PROFILE_SET
    return False


def normalizeProfile(value: object, fallback: Profile = DEFAULT_PROFILE) -> Profile:
    if isinstance(value, ProfileName):
        return value
    if isinstance(value, str) and value in _PROFILE_SET:
        return ProfileName(value)
    return fallback


def normalizeProfiles(values: Iterable[object]) -> list[Profile]:
    return [normalizeProfile(value) for value in values]

File: packages/js/src/core/profiles.ts
export enum ProfileName {
  AFSK_BELL = "afsk-bell",
  AFSK_FIFTH = "afsk-fifth",
  GFSK_FIFTH = "gfsk-fifth",
  MFSK = "mfsk",
}

export type Profile = ProfileName;

export const PROFILE_NAMES: readonly ProfileName[] = Object.freeze([
  ProfileName.AFSK_BELL,
  ProfileName.AFSK_FIFTH,
  ProfileName.GFSK_FIFTH,
  ProfileName.MFSK,
]);

export const DEFAULT_PROFILE: Profile = ProfileName.AFSK_BELL;

const PROFILE_SET = new Set(PROFILE_NAMES);

export function isProfile(value: unknown): value is Profile {
  return typeof value === "string" && PROFILE_SET.has(value as Profile);
}

export function normalizeProfile(value: unknown, fallback: Profile = DEFAULT_PROFILE): Profile {
  return isProfile(value) ? value : fallback;
}
